File: web/src/lib/services/s3LocalTest.ts
import { logger } from '../logger';

const TAG = 's3LocalTest';
const DEFAULT_TIMEOUT_MS = 15_000;

/** User-Agent for direct S3 presigned requests (some providers e.g. CSTCloud Data Capsule validate UA). */
export const S3_COMPATIBLE_USER_AGENT = 'ShrimpSend/1.0 S3Compat';

export function sanitizePresignedUrlForLog(url: string): string {
  try {
    const u = new URL(url);
    u.search = '';
    return u.toString();
  } catch {
    return url.split('?')[0] ?? url;
  }
}

export function presignQueryParam(url: string, key: string): string | null {
  try {
    return new URL(url).searchParams.get(key);
  } catch {
    return null;
  }
}

export function logPresignedUrlSummary(url: string, bucket?: string): void {
  const sanitized = sanitizePresignedUrlForLog(url);
  const signedHeaders = presignQueryParam(url, 'X-Amz-SignedHeaders');
  const credential = presignQueryParam(url, 'X-Amz-Credential');
  let credentialScope: string | null = credential;
  if (credential?.includes('/')) {
    credentialScope = credential.split('/').slice(1).join('/');
  }
  let pathStyleLikely: boolean | null = null;
  if (bucket) {
    try {
      pathStyleLikely = new URL(url).pathname.includes(`/${bucket}`);
    } catch {
      pathStyleLikely = null;
    }
  }
  let host = '';
  try {
    host = new URL(url).host;
  } catch {
    host = '';
  }
  logger.info(
    TAG,
    'presigned url host=',
    host,
    'pathStyleLikely=',
    pathStyleLikely,
    'signedHeaders=',
    signedHeaders,
    'credentialScope=',
    credentialScope,
    'sanitized=',
    sanitized,
  );
}

/**
 * Probe S3 reachability from the client network using a server-issued presigned URL.
 */
export async function headPresignedUrl(
  url: string,
  timeoutMs = DEFAULT_TIMEOUT_MS,
  userAgent?: string,
): Promise<void> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  const ua = userAgent?.trim() || S3_COMPATIBLE_USER_AGENT;
  try {
    const res = await fetch(url, {
      method: 'HEAD',
      signal: controller.signal,
      headers: { 'User-Agent': ua },
    });
    if (res.status >= 200 && res.status < 300) return;

    const sanitized = sanitizePresignedUrlForLog(url);
    let bodySnippet = '';
    try {
      const text = await res.text();
      bodySnippet = text.length > 500 ? text.slice(0, 500) : text;
    } catch {
      bodySnippet = '';
    }
    logger.warn(
      TAG,
      'HEAD failed status=',
      res.status,
      'url=',
      sanitized,
      'x-amz-error-code=',
      res.headers.get('x-amz-error-code'),
      'x-amz-error-message=',
      res.headers.get('x-amz-error-message'),
      'content-type=',
      res.headers.get('content-type'),
      'body=',
      bodySnippet,
    );
    throw new Error(`S3 HEAD failed (HTTP ${res.status})`);
  } catch (e) {
    if (e instanceof DOMException && e.name === 'AbortError') {
      throw new Error('S3 connection timed out');
    }
    throw e;
  } finally {
    clearTimeout(timer);
  }
}

/** Remove legacy localStorage entry that stored AK/SK before client-side cache removal. */
export function clearLegacyS3ConfigCache(): void {
  try {
    localStorage.removeItem('s3_config_local');
  } catch {
    // ignore
  }
}

File: app/lib/services/s3_local_test.dart
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

import '../logger.dart';

const _legacyStorageKey = 's3_config_local';

/// User-Agent for direct S3 presigned requests (some providers e.g. CSTCloud Data Capsule validate UA).
const kS3CompatibleUserAgent = 'ShrimpSend/1.0 S3Compat';

/// Remove legacy SharedPreferences entry that stored AK/SK before cache removal.
Future<void> clearLegacyS3ConfigCache() async {
  final prefs = await SharedPreferences.getInstance();
  await prefs.remove(_legacyStorageKey);
}

String sanitizePresignedUrlForLog(String url) {
  final uri = Uri.tryParse(url);
  if (uri == null) return url.split('?').first;
  return uri.replace(queryParameters: {}).toString();
}

String? presignQueryParam(String url, String key) {
  return Uri.tryParse(url)?.queryParameters[key];
}

void logPresignedUrlSummary(String url, {String? bucket}) {
  final sanitized = sanitizePresignedUrlForLog(url);
  final signedHeaders = presignQueryParam(url, 'X-Amz-SignedHeaders');
  final credential = presignQueryParam(url, 'X-Amz-Credential');
  final pathStyleHint = bucket != null && bucket.isNotEmpty
      ? sanitized.contains('/$bucket')
      : null;
  logApi.info(
    's3 presigned url host=${Uri.tryParse(url)?.host} '
    'pathStyleLikely=$pathStyleHint signedHeaders=$signedHeaders '
    'credentialScope=${credential != null && credential.contains('/') ? credential.split('/').skip(1).join('/') : credential}',
  );
}

Future<void> headPresignedUrl(
  String url, {
  Duration timeout = const Duration(seconds: 15),
  String? userAgent,
}) async {
  final ua = (userAgent != null && userAgent.isNotEmpty)
      ? userAgent
      : kS3CompatibleUserAgent;
  final response = await http
      .head(
        Uri.parse(url),
        headers: {'User-Agent': ua},
      )
      .timeout(timeout);
  if (response.statusCode >= 200 && response.statusCode < 300) return;

  final sanitized = sanitizePresignedUrlForLog(url);
  final bodySnippet = response.body.length > 500
      ? response.body.substring(0, 500)
      : response.body;
  logApi.warning(
    's3 HEAD failed status=${response.statusCode} url=$sanitized '
    'x-amz-error-code=${response.headers['x-amz-error-code']} '
    'x-amz-error-message=${response.headers['x-amz-error-message']} '
    'content-type=${response.headers['content-type']} '
    'body=$bodySnippet',
  );
  throw Exception('S3 HEAD failed (HTTP ${response.statusCode})');
}
